Implement a --format flag on the info command

WilliamMcCumstie opened the pull request:
This change required a solution to the following problems:

1. The output's columns needed to be determined dynamically from a `--format` flag
2. The `NodeHost` options requires each line to represent a single node, and
3. Eventually all the columns will be used in arbitrary combinations

The first problem is fairly straight forward and required making a `Lister` object. Then each column is dynamically "registered" onto the Lister. Otherwise the table is generated in the same manner.

Reconciling problem 2 and 3 proved to be more challenging. Previously the table would show a "pseudo partition" (aka `PartitionProxy`) which grouped the `nodes` into `partitions` by their `state`. This means partitions are duplicated in the output if they contain nodes in multiple states. This needs to be maintained whenever `State` is specified.

However the `NodeHost` option causes each line in the output to represent a single `Node`. Ensure all the outputting options works for both a `Node` and a "pseudo partition" proved to be a herculean task and instead has been side stepped. Instead using `NodeHost` causes each `node` to be grouped into a "pseudo partition" by itself. This essentially changes the output to be on a per `node` basis whilst avoiding re-implementing the table.

WilliamMcCumstie committed:
Add the --format flag to the info command

WilliamMcCumstie committed:
Upgrade info output to be dynamic

This will eventually allow the table columns to be selected.
It will also be able to list nodes as if it was a group

WilliamMcCumstie committed:
Regroup partitions/nodes by hostname or state

WilliamMcCumstie committed:
Configure the list output according to --format

WilliamMcCumstie committed:
Add some additional options to --format

benarmston commented on lib/flight_scheduler/commands/info.rb:
```diff
@@ -28,43 +28,144 @@
+
+        def register_cpus
+          register_column(header: 'CPUS') do |p|
+            p.nodes.reduce(0) { |sum, n| sum += n.cpus.to_i }
```
This should be the number of CPUs per node, not the total number of CPUs.  If the nodes have differing numbers of CPUs it should output "<minimum CPUs>+".  Here is an example from slurm:

```
[flight@gateway1 (ben2) ~]$ grep NodeName /opt/flight/opt/slurm/etc/slurm.conf
NodeName=node01 CPUs=1 RealMemory=1024
NodeName=node02 CPUs=2 RealMemory=2048
[flight@gateway1 (ben2) ~]$ sinfo  --Format NodeHost,CPUs,Memory
HOSTNAMES           CPUS                MEMORY              
node01              1                   1024                
node02              2                   2048                
[flight@gateway1 (ben2) ~]$ sinfo  --Format NodeList,CPUs,Memory
NODELIST            CPUS                MEMORY              
node[01-02]         1+                  1024+               
[flight@gateway1 (ben2) ~]$ 

```

benarmston commented on lib/flight_scheduler/commands/info.rb:
```diff
@@ -28,43 +28,144 @@
+
+        def register_gpus
+          register_column(header: 'GPUS') do |p|
+            p.nodes.reduce(0) { |sum, n| sum += n.gpus.to_i }
```
Similar here to CPUs above.

benarmston commented on lib/flight_scheduler/cli.rb:
```diff
@@ -67,7 +67,18 @@ def self.create_command(name, args_str = '')
     create_command 'info' do |c|
-      c.summary = 'List the available partitions'
+      c.summary = 'List the available partitions and nodes'
+      c.slop.string '-O', '--format', <<~DESC
```
Slurm has both a `-o`/`--format` and a `-O`/`--Format` option.  They are intended to do the same thing, but have different mechanics to achieve that.  I've requested that we copy the `-O`/`--Format` option as I think it is the easier to use and doesn't suffer from the limitations of the other.

What you've implemented is correct (pending some comments below).  However, you've gone with a `-O`/`--format` flag pair which doesn't seem quite right.  I think we should either use `-o`/`--format` or `-O`/`--Format`.  I'm not sure what I prefer here.  Lower case looks nicer, but consistency with slurm is good too.

Would you like to offer a suggestion here?

benarmston commented on lib/flight_scheduler/commands/info.rb:
```diff
@@ -28,43 +28,144 @@
+
+        def register_memory
+          register_column(header: 'MEMORY') do |p|
+            p.nodes.reduce(0) { |sum, n| sum += n.memory.to_i }
```
Similar to CPUs above.

Also we should display the memory in megabytes.  (I'm not interested in supporting machines with less than a meg of memory :-)  As far as memory is concerned, MB is the same as MiB, i.e., 1024*1024.

benarmston reviewed (changes requested):
Looking pretty close.  I've added a few comments I'd like you to look at before this can be merged.

WilliamMcCumstie committed:
Add a comment about duplication of nodes

WilliamMcCumstie committed:
Remove the rogue has-one relationship

It isn't a has-one relationship and never has been

WilliamMcCumstie committed:
Display the minimum CPUs,GPUs, and Memory

WilliamMcCumstie committed:
Implement the alternative formatting method

WilliamMcCumstie committed:
Tweak how the format handles unrecognised fields

WilliamMcCumstie committed:
Remove the PartitionProxy and replace it with helpers

The ParitionProxy assumed that partitions and nodes exist
in a one-to-many relationship, where they actually are in
a many-to-many

For this reason it has been removed and replaced with a set
of helper functions. These helpers easily allow a column
to be defined for a node, partition, or nodes

WilliamMcCumstie committed:
Remove the mutual ex. between node-partition columns

Instead they are sorted into one-one pairs

benarmston commented on lib/flight_scheduler/commands/info.rb:
```diff
@@ -28,43 +28,291 @@
+
+        def value_or_min_plus(*raws)
+          # Ensures everything is an integer
+          # NOTE: Also assumes nil should be interpreted as 0
+          values = raws.map { |v| v.nil? ? 0 : v }
```
Can we default CPUs and Memory to `1`.  The following just looks odd.

```
$ ./bin/scheduler info --Format NodeList,Partition,CPUs,GPUs,Memory
┌───────────────┬───────────┬──────┬──────┬─────────────┐
│ NODELIST      │ PARTITION │ CPUS │ GPUS │ MEMORY (MB) │
├───────────────┼───────────┼──────┼──────┼─────────────┤
│ node01,node02 │ all       │ 0+   │ 0+   │ 0+          │
└───────────────┴───────────┴──────┴──────┴─────────────┘
```

I'm pretty certain that `node02` has at least one CPU and 1MB of memory.  I suppose that `GPU`s could be 0, but not CPU or memory.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler/pull/19/commits/1fc828da8494403b9cb36f8d76f4a835783bf3df

benarmston commented on lib/flight_scheduler/commands/info.rb:
```diff
@@ -28,43 +28,291 @@
+
+        def register_memory
+          register_nodes_column(header: 'MEMORY (MB)') do |nodes|
+            value_or_min_plus(*nodes.map(&:gpus)) do |value|
+              # Convert the memory into MB
+              sprintf('%.2f', value.fdiv(1048576))
```
This should be `&:memory` not `&:gpus`.

Also, lets format it as `%d`.  That way running `free -m` and `scheduler info` both agree.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler/pull/19/commits/1fc828da8494403b9cb36f8d76f4a835783bf3df

benarmston reviewed (changes requested):
We seem to have had quite a miscommunication with the `--format`/`--Format` issue.

I had originally requested two things:

 1.  A single pair of options implementing something similar to slurm's `-O`/`--Format`.
 2. That the options be named `-O`/`--Format`.

Your original implementation was a good first pass at (1), but had named the options `-O`/`--format`, instead of `-O`/`--Format`.

That lead me to wonder whether (2) was the right name for the options.  I wasn't requesting that you also implement additional behaviour similar to slurm's `-o`/`--format`.  Precisely because of all of its wierdo idiosyncrasies.

As we have both now, let's keep them both.

WilliamMcCumstie committed:
Tweak the info command default and fix memory bug

WilliamMcCumstie commented:
Yeah sorry, I miss read the original comment about `--format` and `--Format`. Had implemented the `case` statement before I had realised.

However I still needed to fix the coupling issue between `Partition` and `NodeHost`, so the change got rolled into that work.

benarmston commented:
This looks good to me now.
